notebooks/search/tokenization.ipynb

```diff
@@ -0,0 +1,322 @@
+       " 'later, napoleon and his pigs secretly revise some commandments to clear them of accusations of law - breaking ( such as \" no animal shall drink alcohol \" having \" to excess \" appended to it and \" no animal shall sleep in a bed \" with \" with sheets \" added to it ). the changed commandments are as follows, with the changes bolded : * 4 no animal shall sleep in a bed with sheets. * 5 no animal shall drink alcohol to excess. * 6 no animal shall kill any other animal without cause. eventually these are replaced with the maxims, \" all animals are equal, but some animals are more equal than others \", and \" four legs good, two legs better! \" as the pigs become more human. this is an ironic twist to the original purpose of the seven commandments, which were supposed to keep order within animal farm by uniting the animals together against the humans, and prevent animals from following the humans\\'evil habits. through the revision of the commandments, orwell demonstrates how simply political dogma can be turned into malleable propaganda.']"
+      ]
+     },
+     "execution_count": 10,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "tokens = bert_tokenizer.encode(long_text)[1:-1]  # exclude special tokens at the beginning and end\n",
+    "chunked = [\n",
+    "    bert_tokenizer.decode(tokens_chunk)\n",
```

Comment:
Encoding the text involves a step that normalises the text to Unicode NFD form and then strips non spacing marks (Unicode category Mn).  

Decoding does a reverse look up on the BERT vocabulary file and joins the compound element (those starting with `##`). 

Decoding isn't perfect as those non spacing marks are lost and so may appear different to the original text.

This might be a level of detail too far for most readers but it's worth mentioning. Maybe something like:

"Tokenisation involves a normalisation step and strips non spacing marks. If decoding is implemented as a reverse lookup from token Ids to vocabulary entries those stripped marks will not be recovered resulting in decoded text that could slightly different to the original."

Again I'm not sure we need this level of detail

Reply:
Interesting point. This seems to only be a problem for BERT though. The E5 tokenizer preserves the nonspacing marks:
```py
latvian_name = "Da\u0305vis"
print(latvian_name)
print(bert_tokenizer.decode(bert_tokenizer.encode(latvian_name)))
print(e5_tokenizer.decode(e5_tokenizer.encode(latvian_name)))
```
```
Da̅vis
[CLS] davis [SEP]
<s> Da̅vis</s>
```
If we recommend using ELSER=BERT for English-only and E5 for multi-lingual, I lean towards omitting this point here.